### PR TITLE
Support for OGB Hetero-Graphs

### PR DESCRIPTION
**Pull request by Dsantra92**
Handling Hetero Graphs in OGBDatasets.

### CODE REVIEW

**Comment by CarloLucibello** [comment]
Related to #139

**Comment by CarloLucibello** [comment]
what's the status of this? on which specific ogb dataset are you testing?

**Comment by Dsantra92** [comment]
Testing it on `ogbn-mag`. Edge data section and docs is left. Completing by EOD.

**Comment by Dsantra92** [comment]
Not adding training indexes in this pr. Training splits will be handled as per the [torch issue](https://github.com/JuliaML/MLDatasets.jl/issues/107#issuecomment-1154163832). We also need more documentation for HeteroGraphs (and maybe a tutorial). Not sure why docs are failing :)

File: src/graph.jl [hunk 1]
```diff
@@ -115,10 +116,89 @@ function Base.show(io::IO, ::MIME"text/plain", d::Graph)
...
+struct HeteroGraph <: AbstractGraph
+    num_nodes::Dict{String, Int}
+    num_edges::Dict{Tuple{String, String, String}, Int}
+    edge_indices::Dict{Tuple{String, String, String}, Tuple{Vector{Int}, Vector{Int}}}
+    node_data
+    edge_data
+end
+
+function HeteroGraph(;
+    num_nodes::Union{Dict{String, Int}, Nothing}=nothing,
```
Comment (CarloLucibello) [inline src/graph.jl:156]: do we need `nothing`? can't we just use `Dict{String, Int}()` as default?

**Comment by CarloLucibello** [comment]
Is ` ogbn-mag` big? can we add it to the test set? If not in the CI tests we should at least add it to the local tests

File: src/graph.jl (hunk 1 above)
Comment (Dsantra92) [reply src/graph.jl:156]: Assigning `num_nodes` `Dict{String, Int}()` might indicate that there is no node in the Graph. Whereas `nothing` indicates that `num_nodes` information is not provided but can be inferred from other values.

File: test/datasets/graphs.jl [hunk 2]
```diff
@@ -122,3 +122,26 @@ end
...
+            @test g.num_nodes[node_type] == size(features)[end]
+        end
+    end
+    # check if the number of edges are consistent
+    for edge_data in g.edge_data
+        for (edge_type, features) in edge_data
+            @test g.num_edges[edge_type] == size(features)[end]
+        end
+    end
+end
```
Comment (CarloLucibello) [inline test/datasets/graphs.jl:147]: can you test also the sizes and num_nodes / num_edges specific to this dataset?

**Comment by CarloLucibello** [comment]
`node_data` should be such that `node_data[node_type]` is a dict or a named tuples. Similar applies to edge_data.

File: src/graph.jl [hunk 3]
```diff
@@ -115,10 +116,89 @@ function Base.show(io::IO, ::MIME"text/plain", d::Graph)
...
+- `edge_data`: edge-related data. Can be `nothing`, a named tuple of arrays or a dictionary of arrays.
+            The array's last dimension size should be equal to the number of edges.
+            Default `nothing`.
+"""
+struct HeteroGraph <: AbstractGraph
+    num_nodes::Dict{String, Int}
+    num_edges::Dict{Tuple{String, String, String}, Int}
+    edge_indices::Dict{Tuple{String, String, String}, Tuple{Vector{Int}, Vector{Int}}}
+    node_data
+    edge_data
```
Comment (CarloLucibello) [inline src/graph.jl:152]: should we also add the fields `node_types` and `edge_types`?

Comment (Dsantra92) [reply src/graph.jl:152]: that would also be nice to have

File: test/datasets/graphs.jl (hunk 2 above)
Comment (Dsantra92) [reply test/datasets/graphs.jl:174]: Done!!

**Pull request merged**